### PR TITLE
SQL: jdbc debugging enhancement

### PR DESCRIPTION
**Pull request by astefan**
This change adds a buffered debug output option that will flush the output printer after each debug message when enabled (disabled by default). Also, upon connection time and debug classes initialization, an additional debug log message is generated that contains information about OS, JVM and default JVM timezone setting.

### CODE REVIEW

**Review by matriv — APPROVED** [review]

File: x-pack/plugin/sql/jdbc/src/test/java/org/elasticsearch/xpack/sql/jdbc/JdbcConfigurationTests.java [hunk 1]
```diff
@@ -75,6 +75,23 @@ public void testDebugOut() throws Exception {
         assertThat(ci.debugOut(), is("jdbc.out"));
     }
 
+    public void testDebugBuffered() throws Exception {
+        JdbcConfiguration ci = ci("jdbc:es://a:1/?debug=true&debug.buffered=false");
```
Comment (costin) [inline x-pack/plugin/sql/jdbc/src/test/java/org/elasticsearch/xpack/sql/jdbc/JdbcConfigurationTests.java]: I think `debug.flushAlways` or `debug.noBuffer` is more meaningful. I also like the consistency of the options - by default they are false and need to be turn on through true.
That is the default (`false`) is to always buffer but true would yield an immediate flush.

File: x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/DebugLog.java [hunk 2]
```diff
@@ -17,9 +17,11 @@
     private static final String HEADER = "%tF/%tT.%tL - ";
 
     final PrintWriter print;
+    private boolean debugBuffered;
```
Comment (costin) [inline x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/DebugLog.java]: No need to keep the prefix `debug` as this is a `DebugXXX` class.

File: x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/Debug.java [hunk 3]
```diff
@@ -113,17 +113,16 @@ private static DebugLog logger(JdbcConfiguration info, PrintWriter managedPrinte
             synchronized (Debug.class) {
                 log = OUTPUT_MANAGED.get(managedPrinter);
                 if (log == null) {
-                    log = new DebugLog(managedPrinter);
+                    log = new DebugLog(managedPrinter, info.debugBuffered());
                     OUTPUT_MANAGED.put(managedPrinter, log);
                 }
-                return log;
```
Comment (costin) [inline x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/Debug.java]: The method changes make it more complicated and hard to follow - there's this `log ==null` being repeated which begs the question, what happens it if it initialized?
I think the method is good as it is - it creates a logger or returns one that already exists. Either wrap this method in another one that calls `logSystemInfo` (rename the existing method into something like createLogger and then logger calls `createLogger().logSystemInfo` or wrap that at the consumer site - inside `proxy`.

However currently the systemInfo is being called per each `Connection` which I think is excessive - we want the system information to be once per log, at the beginning.
In which case, the system info should called when a new log is created, essentially after each `new DebugLog`.
To keep things incapsulated, instead of calling the constructor, one could call a method `createLog` which internally calls `new DebugLog` and right after calls logSystemInfo.
This way any other initialization that would need to occur, would happen in that method regardless of the actual method taking place and only _once_ per logger.
It might make sense to call `flush` after logging the system info.

**Review by costin — COMMENTED** [review]
Left a few comments regarding initialization.

**Comment by astefan** [comment]
@costin very valid points. Pushed an update.

**Review by costin — APPROVED** [review]
LGTM

**Review by matriv — APPROVED** [review]
LGTM, nice!
